### PR TITLE
feat(ingestion): Phase 15 Wave 1+2 — ingestion-pipeline reliability

### PR DESCRIPTION
**Pull request by cemililik**
## Summary

Phase 15 (Ingestion Pipeline Reliability) Wave 1 + Wave 2 — closes the silent-failure gap the 2026-05-11 pilot exposed across PDF / DOCX / EPUB / TXT / Markdown ingestion plus the user-facing playground notebook. Targeted for **v0.6.0**.

Also bundles the v0.5.7 review-round-5 follow-up (`b3d6dcb`) that was sitting on `development` after the v0.5.7 tag.

## Wave 1 — silent-failure remediation (10 tasks)

- **Task 1** — `_strip_repeating_page_lines` now inspects the **top-3 / bottom-3 rows** per page instead of the strict outermost row. Variable-outer-line + constant-deeper-line corpora (the audit §1.1 trap) are dedup'd in a single pass; a second pass via `strip_paragraph_packed_headers` mops up survivors after paragraph packing.
- **Task 2** — new `forgelm/_script_sanity.py` runs a per-file Unicode-block check after every extract (`tr` / `en` / `de` / `fr` / `es` / `it` / `pt` with discrete diacritic allow-lists). 1.5 % threshold; emits `script_sanity_summary`.
- **Task 3** — new `forgelm/_pypdf_normalise.py` maps audit-measured fallback glyphs (`ø Õ ú ÷ ࡟`) → `İ ı ş ğ •`. `forgelm doctor` verifies via the new `pypdf_normalise.turkish` diagnostic.
- **Task 4** — end-of-run quality pre-signal (alpha-ratio + weird-char ratio + repeated-line ratio); emits `[WARN] N/M chunks below ingestion quality threshold`.
- **Task 5** — `forgelm audit --quality-filter` is **default-ON** in v0.6.0; new `--no-quality-filter` companion preserves the pre-v0.6.0 opt-in semantics.
- **Task 6** — DOCX `_extract_docx` reads `doc.sections[i].header.paragraphs` + `.footer.paragraphs` and subtracts those lines from the body.
- **Task 7** — EPUB iterates `book.spine` (reading order) + filters `nav` / `cover` / `copyright` / `colophon` / `titlepage` / `frontmatter` items.
- **Task 8** — TXT `encoding="utf-8-sig"` strips BOM; new `_extract_markdown` strips `---\n…\n---\n` YAML frontmatter.
- **Task 9** — `notebooks/ingestion_playground.ipynb` Cells 5 / 8 / 9 / 10 rewritten: token-aware knobs, quality-filter blurb, explicit `--quality-filter`, `quality_summary` pretty-print.
- **Task 10** — `tests/fixtures/ingestion/` carries TXT-BOM + MD-frontmatter + mixed-directory fixtures; PDF / DOCX / EPUB fixtures synthesised in-test via `pypdf.PdfWriter` / `python-docx` / `ebooklib`.

## Wave 2 — strong nice-to-have (5 tasks)

- **Task 11** — `--strip-pattern REGEX` with ReDoS guard (rejects `(a+)+b` shapes + `.*?` + back-reference under DOTALL; 5-second SIGALRM budget on POSIX).
- **Task 12** — `--page-range START-END` PDF flag; new `IngestParameterError(ValueError)` propagates through the per-file soft-fail catch to `EXIT_CONFIG_ERROR (1)`.
- **Task 13** — front-matter / back-matter heuristic (alpha < 0.45 + underscore > 0.10 + ≥ 5 page-numbers); default-ON in v0.6.0; opt-out via `--keep-frontmatter`.
- **Task 14** — `--strip-urls {keep,mask,strip}`; independent of `--all-mask` (URL handling is a content-shape decision, not a GDPR redaction).
- **Task 15** — multi-column PDF detection (warning only) via pypdf `visitor_text` callback sampling.

## Schema additions (all additive — no pre-Phase-15 key renamed)

`IngestionResult`: `pdf_paragraph_packed_lines_stripped`, `script_sanity_triggered`, `strip_pattern_substitutions`, `urls_handled`, `frontmatter_pages_dropped`.

`notes_structured`: same additions + `script_sanity_summary`, `quality_presignal`, `frontmatter_pages_dropped`.

## Test plan

- [x] `ruff format .` + `ruff check .` — clean
- [x] `pytest tests/` — **1728 passed, 14 skipped (optional extras), 0 failed**
- [x] 75 new Phase 15 tests across 4 modules:
  - `tests/test_pypdf_normalise.py` (16)
  - `tests/test_script_sanity.py` (16)
  - `tests/test_strip_pattern.py` (11)
  - `tests/test_ingestion_reliability.py` (36, organised by task)
- [x] `tools/check_bilingual_parity.py --strict` — 46 doc pairs green
- [x] `tools/check_anchor_resolution.py --strict` — 269 files green
- [x] `tools/check_cli_help_consistency.py --strict` — 443 invocations green
- [x] `tools/check_wizard_defaults_sync.py` — schema/JSON sync green
- [x] `tools/check_no_analysis_refs.py` — no public-tree leaks
- [x] `tools/check_no_unguarded_sys_modules_pop.py` — 167 files green
- [x] `forgelm doctor` — `pypdf_normalise.turkish: pass` row renders correctly
- [x] `forgelm ingest --help` — all 11 new flags render with prose help

## Documentation

- `CHANGELOG.md` `[Unreleased]` block
- `docs/guides/ingestion.md` + `-tr.md` — new Phase 15 sections after the page-header dedup block
- `docs/guides/data_audit.md` + `-tr.md` — `--quality-filter` default-ON note
- `docs/reference/data_ingestion_architecture.md` + `-tr.md` — window-based dedup explanation + Phase 15 limitations recap
- `docs/roadmap.md` — Phase 15 row: 📋 Planned → 🚧 In review
- `docs/roadmap/phase-15-ingestion-reliability.md` — Tasks 1-15 checked `[x]`

## Surface coverage check

- ✅ **CLI** — 11 new `ingest` flags + audit `--quality-filter` BooleanOptionalAction
- ✅ **Library API** — `ingest_path()` 11 new kwargs (all optional, back-compat defaults)
- ✅ **Notebook** — `ingestion_playground.ipynb` Tasks 9 acceptance
- ✅ **Doctor** — new `pypdf_normalise.turkish` diagnostic
- ⏭️ **Wizard** — untouched (wizard does not generate `forgelm ingest` invocations)
- ⏭️ **`config_template.yaml`** — untouched (ingestion is CLI-only, never YAML-configurable)

## Reviewer hot spots

1. **Window correctness** in `_pop_windowed_edges` — a line appearing in both `repeating_firsts` and `repeating_lasts` must not be popped twice.
2. **Discrete diacritic allow-lists** in `_script_sanity._LANGUAGE_BLOCKS` — verify no legitimate characters are missing per supported language.
3. **`_TURKISH_MULTI` ordering** — does `"ö " → "Ğ"` collide with legitimate Turkish text ending with `ö`-then-space?
4. **`_QualitySamplingFileHandle.write`** — per-chunk `json.loads` cost on 100K-row corpora.
5. **`_check_unbounded_quantifier_sequence`** heuristic — what ReDoS shapes does it miss? SIGALRM is the safety net.
6. **`_is_frontmatter_page` false positives** — a glossary entry with 5 cross-references could trip the heuristic.
7. **`_maybe_warn_multi_column`** sampling — only first 3 pages; corpora with mid-document 2-column sections are silent.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Language-aware script sanity checks, Turkish glyph normalization, operator regex stripping with safety guard, EPUB/DOCX/TXT/Markdown extraction improvements, PDF page-range/frontmatter heuristics, multi-column PDF warning, and additional ingestion CLI controls.
* **Changed**
  * Audit quality-filter now ON by default (opt-out available); ingestion emits an end-of-run quality presignal by default (opt-out).
* **Documentation**
  * Guides, CLI reference, roadmap, and JSON output docs updated to cover Phase 15 fields and flags.
* **Tests**
  * New regression and unit tests plus fixtures covering Phase 15 behaviors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/cemililik/ForgeLM/pull/49)

### CODE REVIEW

File: tests/test_ingestion_reliability.py [hunk 1]
```diff
@@ -0,0 +1,896 @@
...
+
+        This is a Python-level mirror of the notebook's runtime check. The
+        notebook raises ValueError if exactly one of CHUNK_TOKENS / TOKENIZER
+        is set; we re-derive the same predicate here so a regression in the
+        cell logic is caught by the test suite (the notebook itself is not
+        executed in pytest).
+        """
+        for chunk_tokens, tokenizer in [(512, None), (None, "Qwen/Qwen2.5-7B")]:
+            count = sum(1 for v in (chunk_tokens, tokenizer) if v)
+            assert count == 1, "partial token-mode is the very case the cell rejects"
```
[the remark this answers is not in the document]
Comment (cemililik) [reply tests/test_ingestion_reliability.py]: Resolved in commit 28f4010. The test now parses the actual notebook JSON and asserts Cell 5 carries both the partial-token-mode predicate string and the `raise ValueError` call — a regression deleting the cell-side safety net trips the test instead of silently passing.

File: forgelm/ingestion.py [hunk 2]
```diff
@@ -236,11 +603,237 @@ def _read_pdf_pages(reader: Any, path: Path) -> List[str]:
...
+    if not non_ws:
+        return False
+    alpha_ratio = sum(1 for c in non_ws if c.isalpha()) / len(non_ws)
+    underscore_ratio = text.count("_") / len(text)
+    page_num_hits = len(_FRONTMATTER_PAGE_NUM_PATTERN.findall(text))
+    return (
+        alpha_ratio < _FRONTMATTER_ALPHA_RATIO_MAX
+        and underscore_ratio > _FRONTMATTER_UNDERSCORE_RATIO_MIN
+        and page_num_hits >= 5
+    )
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/ingestion.py:763]: Resolved in commit 28f4010. `_FRONTMATTER_LEADER_RUN_PATTERN = re.compile(r"[._]{3,}")` now counts BOTH underscore- AND dot-runs of length ≥ 3 and denominates the ratio over non-whitespace characters (parity with `alpha_ratio`). Constant renamed `_FRONTMATTER_UNDERSCORE_RATIO_MIN` → `_FRONTMATTER_LEADER_RATIO_MIN`.

File: forgelm/ingestion.py [hunk 3]
```diff
@@ -166,22 +368,162 @@ def _strip_repeating_page_lines(pages: List[str]) -> Tuple[List[str], int]:
...
+            # normaliser's API.
+            substitutions = sum(1 for a, b in zip(text, normalised) if a != b) + abs(len(normalised) - len(text))
+            logger.info(
+                "Normalisation profile %r applied %d substitution(s) on '%s'. "
+                "Pass --no-normalise-unicode or --normalise-profile none to disable.",
+                ctx.normalise_profile,
+                substitutions,
+                file_path,
+            )
+        text = normalised
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/ingestion.py:538]: Resolved in commit 28f4010. New `forgelm._pypdf_normalise.count_substitutions` helper returns the exact per-rule tally (`text.count(src)` per multi-char rule first, then single-char). ingestion.py now reads from that instead of the zip-diff heuristic.

File: forgelm/_strip_pattern.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    finally:
+        signal.alarm(0)
+        signal.signal(signal.SIGALRM, previous_handler)
+        if previous_remaining > 0:
+            # Honour an outer caller's alarm budget (same protocol the
+            # reverse-pii ReDoS guard uses). Round up partial seconds so
+            # the outer budget is preserved, never extended.
+            elapsed = time.monotonic() - started
+            remaining = max(int(previous_remaining - elapsed) + 1, 1)
+            signal.alarm(remaining)
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/_strip_pattern.py:425]: Resolved in commit 28f4010. `_apply_one` now clamps the inner alarm to `min(timeout_s, previous_remaining)` so a nested call can never extend an outer caller's deadline. Outer-restore arithmetic still uses the ORIGINAL `previous_remaining` (not the clamped value) so the parent budget is preserved correctly.

File: docs/usermanuals/tr/data/quality-filter.md [hunk 5]
```diff
@@ -42,8 +42,15 @@ audit:
...
-$ forgelm audit data/ingested.jsonl --quality-filter
+# v0.6.0+: quality-filter DEFAULT-AÇIK; açık flag zararsız.
+$ forgelm audit data/ingested.jsonl
 ✓ 45 satır düşürüldü; data/clean.jsonl yazıldı (12,355 satır)
+
+# v0.6.0 öncesi (veya açıkça belirtmek için), bayrağı geçirin:
+$ forgelm audit data/ingested.jsonl --quality-filter
+
+# CI gate'iniz opt-in semantik'e bağlıysa yeni default'tan opt-out edin:
+$ forgelm audit data/ingested.jsonl --no-quality-filter
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/usermanuals/tr/data/quality-filter.md]: Resolved in commit 28f4010. The TR mirror's shell block was reverted to English-only comments + sample output per CLAUDE.md ("Code, CLI output, logs, config keys are English only"); surrounding prose stays Turkish.

File: docs/usermanuals/en/reference/cli.md [hunk 6]
```diff
@@ -108,10 +108,20 @@ $ forgelm ingest INPUT_PATH \
...
+    [--page-range START-END] [--keep-frontmatter] \
+    [--strip-urls {keep,mask,strip}]
 ```
 
-See [Document Ingestion](#/data/ingestion).
+Phase 15 (v0.6.0) added the `--language-hint`, `--script-sanity-threshold`,
+`--normalise-profile`, `--no-normalise-unicode`, `--no-quality-presignal`,
+`--epub-no-skip-frontmatter`, `--keep-md-frontmatter`, `--strip-pattern`,
+`--strip-pattern-no-timeout`, `--page-range`, `--keep-frontmatter`, and
+`--strip-urls` flags. See [Document Ingestion](#/data/ingestion).
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/usermanuals/en/reference/cli.md:129]: Resolved in commit 28f4010. `docs/usermanuals/{en,tr}/reference/cli.md` ingest synopsis now lists `--output-format {text,json}` with a one-paragraph note pointing operators at the JSON envelope reference.

File: docs/usermanuals/en/data/quality-filter.md [hunk 7]
```diff
@@ -42,8 +42,15 @@ audit:
 ```
 
 ```shell
-$ forgelm audit data/ingested.jsonl --quality-filter
+# v0.6.0+: quality-filter is DEFAULT-ON; the explicit flag is harmless.
+$ forgelm audit data/ingested.jsonl
 ✓ dropped 45 rows; wrote data/clean.jsonl (12,355 rows)
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/usermanuals/en/data/quality-filter.md]: Resolved in commit 28f4010. The default-invocation example no longer claims dropping; it now shows `✓ wrote audit/data_audit_report.json (quality_summary: 45 / 12,400 flagged)` plus an inline comment explaining that drop / write-clean only fires when the YAML keys `audit.quality_filter.drop_flagged` + `write_clean_output` are set.

File: docs/guides/ingestion-tr.md [hunk 8]
```diff
@@ -270,15 +284,177 @@ boyutudur ve `--overlap` ile aynı yarım-pencere üst sınırına tabidir.
...
+  oranı kalibre edilmiş %1.5 eşiğini geçtiğinde (`--script-sanity-threshold`
+  ile ayarlanabilir), bir WARNING tetiklenir + yapılandırılmış
+  `script_sanity_summary` bloğu `notes_structured` içine düşer.
+  Desteklenen diller: `tr`, `en`, `de`, `fr`, `es`, `it`, `pt`
+  (CJK / Arapça Faz 16+'ya ertelendi).
+* **`--normalise-profile {turkish,none}`** çıkarılmış metne
+  dil-spesifik bir glyph normalizasyon tablosu uygular. `turkish`
+  profili (varsayılan) audit'in ölçtüğü artefaktları chunk-write
+  zamanında `İ ı ş ğ •` karşılığına eşler. `--no-normalise-unicode`
+  veya `--normalise-profile none` ile tamamen kapatılabilir.
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/guides/ingestion-tr.md]: Resolved in commit 28f4010. The TR doc now states the default profile is `none`, that `turkish` is auto-derived only when `--language-hint tr` is set, and that an explicit `--normalise-profile turkish` always wins. EN mirror updated in parallel.

File: forgelm/ingestion.py [hunk 9]
```diff
@@ -104,53 +228,131 @@ class IngestionResult:
...
+        while head_keep and tail_drops < window:
+            tail_idx = -1
+            if head_keep[tail_idx] in repeating_lasts:
+                head_keep.pop()
+                stripped += 1
+                tail_drops += 1
+            else:
+                break
+        lines[:] = head_keep
     return stripped
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/ingestion.py:370]: Resolved in commit 81823e7. The tail walk in `_pop_one_page_window` is now symmetric to the head walk — it iterates the bottom-N window in reverse and drops any line whose stripped form is in `repeating_lasts`, regardless of position within the window. New regression test `test_window_based_dedup_strips_deeper_constant_bottom_line` uses your exact counterexample shape (`[..., "FOOTER", "<page-num>"]`) and locks the invariant in alongside the existing top-edge case.

File: tests/test_ingestion_reliability.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cemililik) [reply tests/test_ingestion_reliability.py]: Resolved in commit 28f4010. The test now parses the actual notebook JSON and asserts that Cell 5 carries both (a) the partial-token-mode predicate string `sum(1 for v in (CHUNK_TOKENS, TOKENIZER) if v)` and (b) the `raise ValueError` call with the canonical error message. A regression deleting the cell-side safety net now trips the test instead of silently passing.

File: forgelm/ingestion.py [hunk 10]
```diff
@@ -1431,7 +2250,130 @@ def ingest_path(
...
+        # corrupt the sampler.
+        try:
+            data = json.loads(payload.rstrip("\n"))
+            text = data.get("text", "") if isinstance(data, dict) else ""
+        except (json.JSONDecodeError, ValueError):
+            text = ""
+        if text:
+            self._sampler(text)
+        return self._fh.write(payload)
+
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/ingestion.py]: Resolved in commit 206f157. The `_QualitySamplingFileHandle` wrapper was removed entirely; `_emit_chunk` now takes a `sampler: Optional[Callable[[str], None]]` kwarg and the post-mask chunk text is handed straight to the pre-signal collector — no per-chunk `json.loads` round-trip.

File: forgelm/cli/subcommands/_ingest.py [hunk 11]
```diff
@@ -1,16 +1,66 @@
...
+    if getattr(args, "no_normalise_unicode", False):
+        return "none"
+    profile = getattr(args, "normalise_profile", None)
+    if profile is None:
+        # Library default (turkish) is the right v0.6.0 launch profile.
+        # We pass ``None`` through so library callers can introspect the
+        # default; the ingest_path keyword falls back to the library
+        # constant.
+        return "turkish"
+    return profile
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/cli/subcommands/_ingest.py]: Resolved in commit 206f157. `_resolve_normalise_profile` no longer hardcodes `"turkish"` as the default. The dispatcher now: (1) honours `--no-normalise-unicode` first; (2) honours explicit `--normalise-profile X`; (3) auto-derives `turkish` only when `--language-hint tr` is set; (4) falls back to `none` otherwise. Library `ingest_path` mirrors the same derivation so a Python-level caller without an explicit `normalise_profile` kwarg gets the same behaviour.

File: CHANGELOG.md [hunk 12]
```diff
@@ -4,6 +4,138 @@ All notable changes to ForgeLM are documented here.
...
+- **`docs/roadmap/completed-phases.md::Phase 12` summary** — Tier 1
+  status line claimed a `[ingestion-secrets]` extra "via
+  `detect-secrets` with regex fallback", but that extra was never
+  published; only `[ingestion-pii-ml]` exists in `pyproject.toml`'s
+  extras surface. Reworded to record the historical plan accurately
+  ("`detect-secrets` integration was originally planned as a
+  `[ingestion-secrets]` extra but deferred — only
+  `[ingestion-pii-ml]` ultimately shipped"). Pure docs accuracy fix.
+
 ## [0.5.7] — 2026-05-11
```
[the remark this answers is not in the document]
Comment (cemililik) [reply CHANGELOG.md:139]: Resolved in commit 206f157. The CHANGELOG footer now anchors `[Unreleased]` at `v0.5.7...HEAD` and adds compare URLs for the missing `0.5.6` / `0.5.7` releases.

File: docs/guides/data_audit.md [hunk 13]
```diff
@@ -496,8 +496,10 @@ Hamming-distance cutoff of 3 (≈95 % similarity); ignored when
 `--dedup-method=minhash`. `--dedup-method` (Phase 12) selects the
 near-duplicate engine — `simhash` (default) or `minhash` (needs
 `[ingestion-scale]` extra; `--jaccard-threshold` controls the cutoff,
-default 0.85). `--quality-filter` (Phase 12) opts into the heuristic
-quality scoring. The credential/secrets scan is **always on** — there
+default 0.85). `--quality-filter` (Phase 12, **default-ON from v0.6.0
+per Phase 15 Task 5**) runs the heuristic quality scoring;
+`--no-quality-filter` opts out for operators wanting the pre-v0.6.0
+opt-in semantics. The credential/secrets scan is **always on** — there
 is no flag to disable it.
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/guides/data_audit.md:507]: Resolved in commit 206f157. The "Run it" example and its TR mirror now document the default-on behaviour and the `--no-quality-filter` opt-out companion; the Phase 12 / Phase 15 Task 5 transition is called out explicitly.

File: docs/reference/data_ingestion_architecture-tr.md [hunk 14]
```diff
@@ -224,13 +224,38 @@ notları:
...
+script'leri v0.6.0'da desteklenmemeye devam eder:
+
+- **Çok-kolon** — `_maybe_warn_multi_column` yalnızca WARNING tetikler;
+  okuma sırası pypdf üzerinden hâlâ sol-üst-dan-sağ-alta serileştirilir.
+  Faz 16+ yeni bir `[ingestion-tables]` extra altında camelot-py /
+  pdfplumber fallback'i ekleyebilir.
+- **OCR** — text-layer-tespit retry'ı yok. Audit'in mevcut "Taranmış
+  PDF'lerle çalışma (OCR teslim akışı)" tarifi
+  [`docs/guides/ingestion.md`](../guides/ingestion.md) içindedir;
+  otomatik `ocrmypdf` önerisi Wave 3'e ertelendi (audit §6).
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/reference/data_ingestion_architecture-tr.md:258]: Resolved in commit 206f157. The OCR-recipe link in the TR mirror now points at `../guides/ingestion-tr.md` instead of the EN guide.

File: docs/reference/data_ingestion_architecture.md [hunk 15]
```diff
@@ -222,12 +222,39 @@ Implementation notes from
...
+  so a variable-outer-line corpus (per-chapter title on top, page number
+  on bottom) no longer locks the dedup out from peeling the
+  constant-deeper-line one row deeper. The bug was the loop's
+  *exit condition*, not the outermost-line check — the pre-Phase-15
+  iterator broke as soon as no recurrence was found at the strict
+  outermost position. The fix is a window check, not an extra pass.
+- A **second pass** runs after paragraph packing
+  (`strip_paragraph_packed_headers`) to mop up survivor headers the
+  chunker re-glued mid-block. Surfaces as
+  `pdf_paragraph_packed_lines_stripped` in `notes_structured`.
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/reference/data_ingestion_architecture.md:240]: Resolved in commit 206f157. Both EN + TR architecture mermaid diagrams have been refreshed to show the window-based dedup (`_PDF_EDGE_WINDOW = 3`) and the second-pass `strip_paragraph_packed_headers` node after paragraph packing.

File: docs/roadmap.md [hunk 16]
```diff
@@ -14,7 +14,7 @@
...
-| 📋 Planned | [Phase 15 — Ingestion Pipeline Reliability](roadmap/phase-15-ingestion-reliability.md) | Close silent-failure gap on `forgelm ingest` across PDF / DOCX / EPUB / TXT / Markdown + playground notebook UX alignment; surface the quality signal at ingest time + default-on `--quality-filter` audit; multi-line PDF dedup, Turkish glyph normalisation, DOCX header / footer extraction, EPUB spine + skip, TXT / MD BOM + frontmatter; dual-golden regression fixtures → `v0.6.0` |
+| 🚧 In review | [Phase 15 — Ingestion Pipeline Reliability](roadmap/phase-15-ingestion-reliability.md) | Wave 1 + Wave 2 landed on `development`: window-based multi-line PDF dedup, Turkish glyph normalisation profile, language-aware Unicode-block sanity check, ingest-time quality pre-signal, default-on audit `--quality-filter`, DOCX explicit header/footer subtraction, EPUB spine + nav/cover skip, TXT UTF-8 BOM + MD YAML frontmatter strip, notebook playground alignment, plus Wave 2 `--strip-pattern` (ReDoS-guarded), `--page-range`, front-matter heuristic, `--strip-urls`, multi-column warning → `v0.6.0` |
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/roadmap.md:17]: Resolved in commit 206f157. The status legend now defines `🚧 In review (PR open)` alongside the existing Released / Merged / Planned entries.

File: docs/roadmap/phase-15-ingestion-reliability.md [hunk 17]
```diff
@@ -40,44 +40,44 @@ These splits are estimates, not commitments — task 1's multi-line dedup may su
...
 
-14. [ ] **URL / QR pattern auto-strip option.**
+14. [x] **URL / QR pattern auto-strip option.**
     `--strip-urls` flag: detects URL patterns (http(s)://, www.) per chunk and either replaces with `[URL]` or strips entirely (`--strip-urls=mask|strip|keep`, default `keep`). Useful for exercise-reference QR-noise and academic-paper DOI-noise. Existing PII / secrets mask pipeline integration: applies *after* secrets mask, *before* PII mask, preserving the mask-order discipline. **Interaction with `--all-mask`:** the `--all-mask` shorthand (Phase 12.5) is explicitly *not* extended to include `--strip-urls`. Rationale: PII / secrets are GDPR-grade compliance redactions whose accidental over-application is recoverable (the original text is still in the source PDF); URL stripping is a content-shape decision that operators may want to keep off while still demanding the privacy redactions. The two flag families stay independent and the operator opts into URL handling explicitly. Documented at the CLI help text for both `--all-mask` and `--strip-urls`.
 
-15. [ ] **Multi-column PDF detection.**
+15. [x] **Multi-column PDF detection.**
```
[the remark this answers is not in the document]
Comment (cemililik) [reply docs/roadmap/phase-15-ingestion-reliability.md]: Resolved in commit 206f157. Wave 2 items (11-15) and Wave 3 items (16-21) were renumbered 1-5 / 1-6 respectively so each ordered-list subsection passes markdownlint MD029.

File: forgelm/_strip_pattern.py [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+                # Find the matching opening ``(`` and walk back to see if the
+                # group itself ended on another unbounded quantifier.
+                if _group_ends_with_unbounded_quantifier(pattern, i - 1):
+                    raise StripPatternError(
+                        f"--strip-pattern {pattern!r} contains a nested unbounded "
+                        "quantifier (e.g. `(a+)+`) which is a textbook ReDoS shape. "
+                        "Rewrite with bounded repetition (e.g. `a{1,100}`) or split "
+                        "across multiple narrower --strip-pattern flags."
+                    )
+        i += 1
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/_strip_pattern.py]: Out of scope for round-2/3, intentionally deferred. The structural validator covers the textbook nested shape `(a+)+` / `(\\w+)+`; adjacent quantified groups like `(.+?)(.+)` are a different ReDoS class better handled by the SIGALRM safety net (already documented in `_strip_pattern.py`'s docstring). Adding the adjacent-groups detector risks false positives on benign tandem groups and would require its own test corpus. Tracked as a future iteration if production traffic shows the SIGALRM net firing on this shape.

File: forgelm/_strip_pattern.py [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+# ---------------------------------------------------------------------------
+
+
+def apply_strip_patterns(
+    text: str,
+    patterns: List[Tuple[str, re.Pattern[str]]],
+    *,
+    timeout_s: Optional[int] = DEFAULT_TIMEOUT_S,
+    logger_override: Optional[logging.Logger] = None,
+) -> Tuple[str, int]:
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/_strip_pattern.py:317]: Resolved in commit 206f157. `apply_strip_patterns` now raises `ValueError` on non-positive `timeout_s` instead of silently disabling the alarm; `None` is the explicit "no guard" signal documented in the docstring. Three new test cases pin the invariant.

File: forgelm/cli/_parser.py [hunk 20]
```diff
@@ -294,6 +294,138 @@ def _add_ingest_subcommand(subparsers) -> None:
...
+    p.add_argument(
+        "--script-sanity-threshold",
+        type=_non_negative_float,
+        default=None,
+        metavar="X",
+        help=(
+            "Phase 15 Wave 1: ratio (0.0-1.0) above which the script-sanity check warns "
+            "(default: library calibrated 0.015 = 1.5%%). Operators on noisy mixed-script "
+            "corpora can relax this; values < 1%% start false-positiving on legitimate "
+            "English-with-code-samples Turkish corpora."
```
[the remark this answers is not in the document]
Comment (cemililik) [reply forgelm/cli/_parser.py:320]: False positive. `--script-sanity-threshold` already uses `_non_negative_float` (`forgelm/cli/_argparse_types.py:54`) which raises `ArgumentTypeError` for values outside `[0.0, 1.0]`. The documented range and the validator already agree — no change required.

File: notebooks/ingestion_playground.ipynb [hunk 21]
```diff
@@ -142,7 +155,15 @@
...
     "    parts.append(\"--recursive\")\n",
     "parts += [\"--strategy\", shlex.quote(STRATEGY)]\n",
-    "parts += [\"--chunk-size\", str(CHUNK_SIZE)]\n",
+    "if CHUNK_TOKENS is not None and TOKENIZER is not None:\n",
+    "    # Token-aware mode (Phase 11.5 + Phase 15): the --chunk-size flag is\n",
+    "    # logged as ignored on the ingest side, so we omit it from the command.\n",
+    "    parts += [\"--chunk-tokens\", str(CHUNK_TOKENS)]\n",
+    "    parts += [\"--tokenizer\", shlex.quote(TOKENIZER)]\n",
+    "else:\n",
+    "    parts += [\"--chunk-size\", str(CHUNK_SIZE)]\n",
```
[the remark this answers is not in the document]
Comment (cemililik) [reply notebooks/ingestion_playground.ipynb]: Resolved in commit 206f157. Cell 5 now raises `ValueError` with a clear message when exactly one of `CHUNK_TOKENS` / `TOKENIZER` is truthy; the silent fallback to `--chunk-size` is gone. A dedicated regression test `test_partial_token_mode_rejected_at_notebook_layer` (refined in 28f4010) parses the notebook JSON and asserts the safety net is in place.
